Feature: Game setting to define how industries with neutral stations accept and supply cargo from/to surrounding stations.

PeterN opened the pull request:
Based on #7204 but done by keeping a direct two-way reference between industry and station. This removes associated tile loops to determine where neutral stations are linked.

glx22 commented on src/station_cmd.cpp:
```diff
@@ -526,7 +527,7 @@ CargoArray GetProductionAroundTiles(TileIndex tile, int w, int h, int rad)
  * @param always_accepted bitmask of cargo accepted by houses and headquarters; can be NULL
  * @param st_ind IndustryID associated with Station, if any (e.g. oil rig or INVALID_INDUSTRY)
```
needs to be updated too

nielsmh commented on src/station_func.h:
```diff
@@ -28,7 +29,7 @@ void ShowStationViewWindow(StationID station);
 
 CargoArray GetProductionAroundTiles(TileIndex tile, int w, int h, int rad);
-CargoArray GetAcceptanceAroundTiles(TileIndex tile, int w, int h, int rad, CargoTypes *always_accepted = NULL);
+CargoArray GetAcceptanceAroundTiles(TileIndex tile, int w, int h, int rad, CargoTypes *always_accepted = NULL, const Industry *ind = NULL);
```
This is never used anywhere that isn't a station querying, or a player attempting to build a station, I assume?

PeterN replied:
That is correct.

nielsmh commented:
I definitely like this better than #7204, it avoids propagating the concern into the base tile acceptance query function. And there's a savegame upgrade anyway for the setting, so no real additional complexity for adding another field.

PeterN commented:
@nielsmh Less lines added as well (although many were comments), and less tile iterating, for a tiny memory cost.

PeterN commented:
I am thinking about changing this to store IDs rather than pointers, depending on how the values are used it may make more sense.

PeterN committed:
Feature: Industries with neutral stations (e.g. Oil Rig) only supply/accept cargo to/from their neutral station.

This change is a controlled by a game setting, located under Environment ->
Industries which allows toggling the behaviour. It defaults to enabled.

"Company stations can serve industries with attached neutral stations"

When enabled, industries with attached neutral station (such as Oil Rigs) may
also be served by company-owned stations built nearby. This is the traditional
behaviour.

When disabled, these industries may only be served by their neutral station.
Any nearby company-owned stations won't be able to serve them, nor will the
neutral station serve anything else other than the industry.

glx22 commented on src/saveload/saveload.h:
```diff
@@ -291,6 +291,7 @@ enum SaveLoadVersion : uint16 {
 	SLV_GROUP_LIVERIES,                     ///< 205  PR#7108 Livery storage change and group liveries.
 	SLV_SHIPS_STOP_IN_LOCKS,                ///< 206  PR#7150 Ship/lock movement changes.
 	SLV_FIX_CARGO_MONITOR,                  ///< 207  PR#7175 Cargo monitor data packing fix to support 64 cargotypes.
+	SLV_SERVE_NEUTRAL_INDUSTRIES,           ///< 208  PR#7204 Company stations can serve industries with attached neutral stations.
```
It's now PR#7234

PeterN replied:
Oh yes :-)

PeterN commented:
Thinking about this, I'm tempted to say the default should be the new behviour, with the option there to allow the old 'broken' behaviour.

nielsmh commented:
Any more needs to be done on this, apart from fix saveload.h?

PeterN commented:
I believe it is complete, but as usual more eyes are better. It had bugfix commits which have already been squashed.

michicc commented:
Sorry, saveload conflict.
